crates/valence_server/src/layer/chunk/loaded.rs

```diff
@@ -295,6 +293,72 @@ impl LoadedChunk {
         self.assert_no_changes();
     }
 
+    /// Generates the `MOTION_BLOCKING` heightmap for this chunk, which stores
+    /// the height of the highest non motion-blocking block in each column.
+    ///
+    /// The lowest value of the heightmap is 0, which means that there are no
+    /// motion-blocking blocks in the column. In this case, rain will fall
+    /// through the void and there will be no rain particles.
+    ///
+    /// A value of 1 means that rain particles will appear at the lowest
+    /// possible height y=-64, but note that blocks cannot be placed at y=-65.
+    ///
+    /// We take these two special cases into account by adding a value of 2 to
+    /// our heightmap if we find a motion-blocking block, since
+    /// `self.block_state(x, 0, z)` corresponds to the block at (x, -64, z)
+    /// ingame.
+    #[allow(clippy::needless_range_loop)]
+    fn motion_blocking(&self) -> Vec<Vec<u32>> {
+        let mut heightmap: Vec<Vec<u32>> = vec![vec![0; 16]; 16];
+
+        for z in 0..16 {
+            for x in 0..16 {
+                for y in (0..self.height()).rev() {
+                    if self.block_state(x as u32, y, z as u32).blocks_motion() {
+                        heightmap[z][x] = y + 2;
+                        break;
+                    }
+                }
+            }
+        }
+
+        heightmap
+    }
+
+    /// Encodes a given heightmap into the correct format of the
+    /// `ChunkDataS2c` packet.
+    ///
+    /// The heightmap values are stored in a long array. Each value is encoded
+    /// as a 9-bit unsigned integer, so every long can hold at most seven
+    /// values. The long is padded at the left side with a single zero. Since
+    /// there are 256 values for 256 columns in a chunk, there will be 36
+    /// fully filled longs and one half-filled long with four values. The
+    /// remaining three values in the last long are left unused.
+    ///
+    /// For example, the `WORLD_SURFACE` heightmap in an empty superflat world
+    /// is always 4. The first 36 long values will then be
+    ///
+    /// 0 000000100 000000100 000000100 000000100 000000100 000000100 000000100,
+    ///
+    /// and the last long will be
+    ///
+    /// 0 000000000 000000000 000000000 000000100 000000100 000000100 000000100.
+    fn encode_heightmap(&self, heightmap: Vec<Vec<u32>>) -> Value {
+        let mut encoded: Vec<i64> = vec![0; 37];
```

Comment:
Would be nice if the magic numbers here were computed from local constants to make their derivation more obvious. Specifically, `37`, `7`, `9`.

Reply:
Would it make sense to create a constant such as `CHUNK_LENGTH` for the value 16 or is that value ubiquitous enough?

Reply:
I think it's clear enough what `16` means, but I'm not opposed to it if the code were changed to use it everywhere.